swarm/utils/metrics.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+//NOTE: The metrics package for swarm should align with the general go-ethereum metrics,
+//which uses the `rcrowley/go-metrics` package. The latter has limitations with
+//aggregating and resetting metrics, which sparked some research on the swarm team
+//regarding metrics. Therefore, this swarm metrics package introduces some indirection,
+//which abstracts a metrics API, allowing to use a different metrics library if needed.
+//If ultimately the go-metrics issues are solved,
+//it may be omitted in favor of directly using the `metrics/metrics` package in go-ethereum
+
+package utils
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+
+	graphite "github.com/cyberdelia/go-metrics-graphite"
+	gometrics "github.com/rcrowley/go-metrics"
+)
+
+var (
+	MetricsEnabled = metrics.Enabled
+	bzzAccount     string
+	SwarmRegistry  gometrics.Registry
+)
+
+type MetricsTimer struct {
+	timer  gometrics.Timer
+	bucket string
+}
+
+func SetupMetrics(bzzAccount string) {
+	bzzAccount = bzzAccount[:12]
+	addr, _ := net.ResolveTCPAddr("tcp", "127.0.0.1:2003")
+	SwarmRegistry = gometrics.NewRegistry()
+
+	go graphite.Graphite(
+		SwarmRegistry,
+		5*time.Second,
+		"swarm.node."+bzzAccount,
+		addr)
+}
+
+func Gauge(bucket string, value interface{}) {
+	if metrics.Enabled {
+		g := gometrics.GetOrRegisterGauge(bucket, SwarmRegistry)
+		if val, ok := value.(int); ok {
+			g.Update(int64(val))
+			return
+		}
+		if val, ok := value.(uint64); ok {
+			g.Update(int64(val))
+			return
+		}
+		if val, ok := value.(int64); ok {
+			g.Update(val)
+			return
+		}
+		log.Warn(fmt.Sprintf("Invalid value type for Gauge %v", reflect.TypeOf(value)))
+	}
+}
+
+func Increment(bucket string) {
```

Comment:
Nitpick, but I find `bucket` confusing. We generally measure, or increment `measurements` (or `counters`, but this is just a sub-type) and we time `spans`... I think we should find a better name for this if we stick with the `utils` package.